File: app/src/main/java/com/unitpricecalculator/unit/Units.java
package com.unitpricecalculator.unit;

import androidx.annotation.Nullable;
import com.google.common.base.Function;
import com.google.common.collect.ImmutableList;
import com.squareup.otto.Bus;
import com.unitpricecalculator.currency.Currencies;
import com.unitpricecalculator.events.UnitTypeChangedEvent;
import com.unitpricecalculator.util.prefs.Keys;
import com.unitpricecalculator.util.prefs.Prefs;
import dagger.Reusable;
import java.text.NumberFormat;
import java.util.Currency;
import java.util.HashMap;
import java.util.Map;
import javax.inject.Inject;

/**
 * Collection of utility functions for {@link Unit}.
 */
@Reusable
public final class Units {

  private static final String KEY_CURRENCY = "currency";

  private final Prefs prefs;
  private final Bus bus;

  @Nullable
  private Currency currency;
  @Nullable
  private Function<Double, String> costFormatter;

  @Inject
  Units(Prefs prefs, Bus bus) {
    this.prefs = prefs;
    this.bus = bus;
  }

  private static Map<UnitType, ImmutableList<DefaultUnit>> unitMap = new HashMap<>();

  public ImmutableList<DefaultUnit> getUnitsForType(UnitType unitType) {
    if (unitMap.get(unitType) == null) {
      ImmutableList.Builder<DefaultUnit> list = ImmutableList.builder();
      for (DefaultUnit unit : DefaultUnit.values()) {
        if (unit.getUnitType() == unitType) {
          list.add(unit);
        }
      }
      unitMap.put(unitType, list.build());
    }
    return unitMap.get(unitType);
  }

  public UnitType getCurrentUnitType() {
    return UnitType.valueOf(prefs.getString(Keys.UNIT_TYPE, UnitType.WEIGHT.name()));
  }

  public void setCurrentUnitType(UnitType unitType) {
    prefs.putString(Keys.UNIT_TYPE, unitType.name());
    bus.post(new UnitTypeChangedEvent(unitType));
  }

  public Currency getCurrency() {
    if (currency != null) {
      return currency;
    }

    String savedCurrency = prefs.getString(KEY_CURRENCY);
    if (savedCurrency != null) {
      currency =
          Currencies.parseCurrencySafely(savedCurrency).or(Currencies.getSafeDefaultCurrency());
      costFormatter = null;
      return currency;
    }

    currency = Currencies.getSafeDefaultCurrency();
    costFormatter = null;
    return currency;
  }

  public void setCurrency(Currency currency) {
    this.currency = currency;
    this.costFormatter = null;
    prefs.putString(KEY_CURRENCY, currency.getCurrencyCode());
  }

  public Function<Double, String> getFormatter() {
    if (costFormatter == null) {
      costFormatter = createFormatter();
    }
    return costFormatter;
  }

  private Function<Double, String> createFormatter() {
    NumberFormat lotsOfDigits = NumberFormat.getCurrencyInstance();
    lotsOfDigits.setCurrency(currency);
    lotsOfDigits.setMinimumFractionDigits(2);
    lotsOfDigits.setMaximumFractionDigits(8);

    NumberFormat defaultForCurrency = NumberFormat.getCurrencyInstance();
    defaultForCurrency.setCurrency(currency);

    return input -> {
      String formattedPricePer = lotsOfDigits.format(input);

      String[] parts = formattedPricePer.split("[.,]");
      if (parts.length != 2) {
        return formattedPricePer;
      }

      // Attempt to account for rounding errors and excessive digits by checking for repeating
      // digits after the decimal.
      String decimalPortion = parts[1];
      int longestRunLength = 1;
      int currentRunLength = 1;
      char lastDigit = decimalPortion.charAt(0);
      for (int i = 1; i < decimalPortion.length(); i++) {
        if (decimalPortion.charAt(i) == lastDigit) {
          currentRunLength += 1;
          if (currentRunLength > longestRunLength) {
            longestRunLength = currentRunLength;
          }
        } else {
          currentRunLength = 1;
          lastDigit = decimalPortion.charAt(i);
        }
      }

      if (longestRunLength >= 4) {
        formattedPricePer = defaultForCurrency.format(input);
      }

      return formattedPricePer;
    };
  }
}

File: app/src/main/java/com/unitpricecalculator/unit/Units.kt
package com.unitpricecalculator.unit

import androidx.collection.ArrayMap
import com.google.common.base.Function
import com.squareup.otto.Bus
import com.unitpricecalculator.currency.Currencies
import com.unitpricecalculator.events.UnitTypeChangedEvent
import com.unitpricecalculator.unit.Systems.Companion.includes
import com.unitpricecalculator.unit.UnitType.QUANTITY
import com.unitpricecalculator.util.isIntegral
import com.unitpricecalculator.util.logger.Logger
import com.unitpricecalculator.util.prefs.Keys
import com.unitpricecalculator.util.prefs.Prefs
import dagger.Reusable
import java.text.NumberFormat
import java.util.Currency
import javax.inject.Inject
import kotlin.math.pow

/** Collection of utility functions for [Unit]. */
@Reusable
class Units @Inject internal constructor(
    private val prefs: Prefs,
    private val bus: Bus,
    private val systems: Systems
) {

    private val currencyToFormatter = mutableMapOf<String, CostFormatter>()
    private var _currency: Currency? = null

    var currentUnitType: UnitType
        get() = UnitType.valueOf(prefs.getNonNullString(Keys.UNIT_TYPE, UnitType.WEIGHT.name))
        set(unitType) {
            prefs.putString(Keys.UNIT_TYPE, unitType.name)
            bus.post(UnitTypeChangedEvent(unitType))
        }

    val formatter: CostFormatter
        get() = getCostFormatter()

    fun getCostFormatter(currency: Currency? = null): CostFormatter {
        val effectiveCurrency = currency ?: this.currency
        return currencyToFormatter.getOrPut(effectiveCurrency.symbol) {
            createFormatter(effectiveCurrency)
        }
    }

    fun getUnitsForType(unitType: UnitType): List<DefaultUnit> {
        return unitMap.getOrPut(unitType) {
            DefaultUnit.values().filter { it.unitType == unitType }
        }
    }

    @JvmOverloads
    fun getIncludedUnitsForTypeSorted(
        unitType: UnitType,
        unitToAlwaysInclude: Unit? = null
    ): List<DefaultUnit> {
        val includedSystems = systems.includedSystems
        val preferredOrder = systems.preferredOrder
        return getUnitsForType(unitType)
            .filter {
                it == unitToAlwaysInclude || it.unitType == QUANTITY || includedSystems.includes(it.system)
            }
            .sortedBy {
                if (it.system == System.IMPERIAL) {
                    if (preferredOrder[0] == System.METRIC) {
                        1
                    } else {
                        0
                    }
                } else {
                    preferredOrder.indexOf(it.system)
                }
            }
    }

    var currency: Currency
        get() {
            _currency?.let { return it }

            return (prefs.getString(KEY_CURRENCY)
                ?.let { Currencies.parseCurrencySafely(it).orNull() }
                ?: Currencies.getSafeDefaultCurrency())
                .also {
                    _currency = it
                }
        }
        set(value) {
            this._currency = value
            prefs.putString(KEY_CURRENCY, value.currencyCode)
        }

    private fun createFormatter(currency: Currency): CostFormatter {
        val noFractionFormat = NumberFormat.getCurrencyInstance()
        noFractionFormat.currency = currency
        noFractionFormat.minimumFractionDigits = 0
        noFractionFormat.maximumFractionDigits = 0

        val defaultForCurrency = NumberFormat.getCurrencyInstance()
        defaultForCurrency.currency = currency

        return CostFormatter { input ->
            if (input.isIntegral()) {
                return@CostFormatter noFractionFormat.format(input)
            }

            val lotsOfDigits = NumberFormat.getCurrencyInstance()
            lotsOfDigits.currency = currency
            lotsOfDigits.minimumFractionDigits = maxOf(lotsOfDigits.minimumFractionDigits, 2)

            val firstSignificantFractionIndex = input.firstSignificantFractionIndex
            if (firstSignificantFractionIndex >= 0) {
                lotsOfDigits.maximumFractionDigits =
                    minOf(
                        8, maxOf(
                            lotsOfDigits.maximumFractionDigits,
                            input.firstSignificantFractionIndex + 4
                        )
                    )
            } else {
                lotsOfDigits.maximumFractionDigits = lotsOfDigits.minimumFractionDigits
            }

            lotsOfDigits.format(input)
        }
    }

    private val Double.firstSignificantFractionIndex: Int
        get() = ((0 until 9).firstOrNull { this >= 10.0.pow(-it.toDouble()) } ?: 9) - 1

    fun getDefaultQuantity(): Quantity {
        return getDefaultQuantity(currentUnitType)
    }

    fun getDefaultQuantity(unitType: UnitType): Quantity {
        val amount = prefs.getDouble(getDefaultQuantityAmountKey(unitType))
        val unit = prefs.getString(getDefaultQuantityUnitKey(unitType))
        if (amount == null || unit == null) {
            val targetUnit = getIncludedUnitsForTypeSorted(unitType).first()
            return Quantity(targetUnit.defaultQuantity.toDouble(), targetUnit)
        }

        return Quantity(amount, DefaultUnit.valueOf(unit))
    }

    object DefaultQuantityChangedEvent

    fun setDefaultQuantity(unitType: UnitType, quantity: Quantity) {
        prefs.putDouble(getDefaultQuantityAmountKey(unitType), quantity.amount)
        prefs.putString(getDefaultQuantityUnitKey(unitType), quantity.unit.name)
        bus.post(DefaultQuantityChangedEvent)
        Logger.d("New default quantity: %s, %s", unitType.name, quantity.toString())
    }

    fun interface CostFormatter {
        fun format(cost: Double): String
    }

    companion object {

        private const val KEY_CURRENCY = "currency"

        private val unitMap = ArrayMap<UnitType, List<DefaultUnit>>()

        private fun getDefaultQuantityAmountKey(unitType: UnitType): String {
            return "default_quantity_amount:${unitType.name}"
        }

        private fun getDefaultQuantityUnitKey(unitType: UnitType): String {
            return "default_quantity_unit:${unitType.name}"
        }
    }
}
